src/context.rs

```diff
@@ -957,6 +957,86 @@ impl CDFContext {
     }
   }
 
+  pub fn reset_counts(&mut self) {
+    macro_rules! reset_1d {
+      ($field:expr) => (let l = $field.len(); $field[l-1] = 0; )
```

Comment:
$field.last_mut() would return Option<&mut T> as I understand it. Not sure this would be the code simpler

Reply:
It would be something like this I guess
``` rust
{ let f $field.last_mut().unwrap(); *f = 0; }
``` 

Possibly it makes easier to see what the code is supposed to do.

Reply:
Since the counter of cdf array is located at the last element (except the special case partition_cdf), and each cdf can be defined under multi-dimentional contexts, 1 to 3 here, fbossen original code could be easier to read (at lest to me) to keep consistency over the varying dimensions of array for contexts.